Hatch intake subsystem code

njohnson0000 opened the pull request.

COREProgrammers committed:
Trying to fix some problems

COREProgrammers committed:
One more change

COREProgrammers committed:
Not done

COREProgrammers committed:
Motor added and Pneumatic taken out of the Hatch Intake Subsystem

Julia-Pockat commented on src/HatchIntakeSubsystem.cpp:
```diff
@@ -1,40 +1,54 @@
         SetIntakeOn();
-    }else{
+    } else {
         SetIntakeOff();
```
Right now the controls are set up so that you need to press a couple different buttons in order to operate the hatch intake. Instead of doing this, set it up so that when you only need to press one button while going down, and one button for outtaking the hatch.

Julia-Pockat commented on src/HatchIntakeSubsystem.cpp:
```diff
@@ -44,17 +58,22 @@ void HatchIntakeSubsystem::SetIntakeOff(){
-        m_hatchIntakeSolenoid.Set(frc::DoubleSolenoid::kForward);
-        isDown = true;
+//Actuates the intake up.
+void HatchIntakeSubsystem::HatchActuatorUp(){
```
Rename to SetHatchActuartorUp

Julia-Pockat replied:
I don't think that you really need to separate the up and down functions and the on and off functions. Just make one that can toggle based on the buttons that are pressed.

Julia-Pockat commented on src/HatchIntakeSubsystem.cpp:
```diff
@@ -44,17 +58,22 @@ void HatchIntakeSubsystem::SetIntakeOff(){
+}
+
+//Actuates the intake down.
+void HatchIntakeSubsystem::HatchActuatorDown(){
```
Rename to SetHatchActuatorDown

Julia-Pockat commented on src/HatchIntakeSubsystem.cpp:
```diff
@@ -1,40 +1,54 @@
-    if(operatorJoystick->GetRisingEdge(CORE::COREJoystick::JoystickButton::RIGHT_TRIGGER)){
-        ToggleHatchIntake();
+    if (operatorJoystick->GetRisingEdge(CORE::COREJoystick::JoystickButton::RIGHT_TRIGGER)){
+        if(m_hatchActuator.GetSelectedSensorPosition() == m_actuatorTopLimit.Get()){
```
This code would need to be deadbanded. Right now, the sensor position would need to be exactly equal to the top limit of the actuator in order to set the hatch to the down position, but you would really want a range of values that that the intake would be at to set it down.

Julia-Pockat commented on src/HatchIntakeSubsystem.cpp:
```diff
@@ -1,40 +1,54 @@
     operatorJoystick->RegisterButton(CORE::COREJoystick::JoystickButton::RIGHT_BUTTON);
     operatorJoystick->RegisterButton(CORE::COREJoystick::JoystickButton::RIGHT_TRIGGER);
+    
+    m_hatchIntake.SetNeutralMode(NeutralMode::Coast);
```
I don't understand why you added lines 18 and 19, because it does not seem necessary. Also don't understand why one of them was set to coast and one of them was set to break.

COREProgrammers committed:
Hatch intake changes in response to Julia's comments

COREProgrammers committed:
Added a space between the ) and the {

njohnson0000 commented:
@Julia-Pockat Can you review this code again?

Julia-Pockat commented on src/HatchIntakeSubsystem.cpp:
```diff
@@ -26,32 +26,32 @@ void HatchIntakeSubsystem::robotInit(){
         }
-        if(m_hatchActuator.GetSelectedSensorPosition() <= m_actuatorBottomLimit.Get() + m_hatchActuatorDeadband.Get()){
-            if(GetHatchState()){
+        if(m_hatchActuator.GetSelectedSensorPosition() <= m_actuatorBottomLimit.Get() + m_hatchActuatorDeadband.Get()) {
```
This is not deadbanding the values correctly. In order to deadband you need to make sure that the absolute value of the difference between the desired position and the current position is less than the deadband range.

Julia-Pockat commented on src/HatchIntakeSubsystem.cpp:
```diff
@@ -26,32 +26,32 @@ void HatchIntakeSubsystem::robotInit(){
-    if (operatorJoystick->GetRisingEdge(CORE::COREJoystick::JoystickButton::RIGHT_TRIGGER)){
-        if(m_hatchActuator.GetSelectedSensorPosition() >= m_actuatorTopLimit.Get() - m_hatchActuatorDeadband.Get()){
+    if (operatorJoystick->GetRisingEdge(CORE::COREJoystick::JoystickButton::RIGHT_TRIGGER)) {
+        if(m_hatchActuator.GetSelectedSensorPosition() >= m_actuatorTopLimit.Get() - m_hatchActuatorDeadband.Get()) {
```
See comment below

Julia-Pockat commented on src/HatchIntakeSubsystem.cpp:
```diff
@@ -12,7 +12,7 @@ HatchIntakeSubsystem::HatchIntakeSubsystem() : m_hatchIntake(HATCH_INTAKE_WHEEL_
+void HatchIntakeSubsystem::robotInit() {
     operatorJoystick->RegisterButton(CORE::COREJoystick::JoystickButton::RIGHT_BUTTON);
     operatorJoystick->RegisterButton(CORE::COREJoystick::JoystickButton::RIGHT_TRIGGER);
     
```
I don't understand why you are trying to set the neutral mode of each of these talons, and why you set the modes to two different modes.

Julia-Pockat commented on src/HatchIntakeSubsystem.cpp:
```diff
@@ -26,32 +26,32 @@ void HatchIntakeSubsystem::robotInit(){
-        if(m_hatchActuator.GetSelectedSensorPosition() <= m_actuatorBottomLimit.Get() + m_hatchActuatorDeadband.Get()){
-            if(!GetHatchState()){
+        if(m_hatchActuator.GetSelectedSensorPosition() <= m_actuatorBottomLimit.Get() + m_hatchActuatorDeadband.Get()) {
+            if(!GetHatchState()) {
```
Rename GetHatchState function to describe what the function actually returns.

Julia-Pockat commented on src/HatchIntakeSubsystem.cpp:
```diff
@@ -61,9 +61,9 @@ void HatchIntakeSubsystem::teleop(){
-void HatchIntakeSubsystem::SetIntake(bool setIntakeDirection, bool activation){
-    if(activation){
-        if(setIntakeDirection){
+void HatchIntakeSubsystem::SetIntake(bool setIntakeDirection, bool activation) {
```
Why do you have three functions that handle the actuator movement?

Julia-Pockat replied:
What is activation?

Julia-Pockat replied:
Where is the hatchIntake code for this function?

Julia-Pockat commented on src/HatchIntakeSubsystem.cpp:
```diff
@@ -26,32 +26,32 @@ void HatchIntakeSubsystem::robotInit(){
-    if (operatorJoystick->GetRisingEdge(CORE::COREJoystick::JoystickButton::RIGHT_BUTTON)){
-       if(m_hatchActuator.GetSelectedSensorPosition() >= m_actuatorTopLimit.Get() - m_hatchActuatorDeadband.Get()){
+    if (operatorJoystick->GetRisingEdge(CORE::COREJoystick::JoystickButton::RIGHT_BUTTON)) {
+       if(m_hatchActuator.GetSelectedSensorPosition() >= m_actuatorTopLimit.Get() - m_hatchActuatorDeadband.Get()) {
```
This should be in the function that determines how to set the motors.

Julia-Pockat commented on src/HatchIntakeSubsystem.cpp:
```diff
@@ -26,32 +26,32 @@ void HatchIntakeSubsystem::robotInit(){
     // When the right button is pressed, the intake actuates down and spits out the hatch and actuates up.
-    if (operatorJoystick->GetRisingEdge(CORE::COREJoystick::JoystickButton::RIGHT_BUTTON)){
-       if(m_hatchActuator.GetSelectedSensorPosition() >= m_actuatorTopLimit.Get() - m_hatchActuatorDeadband.Get()){
+    if (operatorJoystick->GetRisingEdge(CORE::COREJoystick::JoystickButton::RIGHT_BUTTON)) {
```
You don't want to be using GetRisingEdge, because you want this to be looped through for as long as you are pressing the button.

COREProgrammers committed:
The response to futher commenting from Julia.

njohnson0000 commented:
@Julia-Pockat can you review this code for hopefully the last time?

COREProgrammers committed:
Continued changes to respond to Julia's comments.

Julia-Pockat commented on src/HatchIntakeSubsystem.cpp:
```diff
@@ -28,27 +28,29 @@ void HatchIntakeSubsystem::teleopInit() {
+// conditionMet is whether the hatch needs to be outtaked or intaked.
 void HatchIntakeSubsystem::SetIntake(bool setIntakeDirection, bool conditionMet) {
     if(setIntakeDirection) {
         if(conditionMet) {
```
Rename conditionMet to something more descriptive

Julia-Pockat commented on src/HatchIntakeSubsystem.cpp:
```diff
@@ -28,27 +28,29 @@ void HatchIntakeSubsystem::teleopInit() {
 
+// setIntakeDirection determines direction, true = in, false = out.
+// conditionMet is whether the hatch needs to be outtaked or intaked.
 void HatchIntakeSubsystem::SetIntake(bool setIntakeDirection, bool conditionMet) {
```
The SetIntake function does not actually set the hatch intake motor. All of your functions are using the hatch actuator, but you want one function that will control both the hatch actuator and the hatch intake.

Julia-Pockat replied:
Fix formatting in this section

Julia-Pockat replied:
SetIntakeDirection needs to be changed to something that describes what you are saying is true or false.

Julia-Pockat commented on src/HatchIntakeSubsystem.cpp:
```diff
@@ -65,24 +67,23 @@ void HatchIntakeSubsystem::SetIntake(bool setIntakeDirection, bool conditionMet)
-    } else {
-        m_hatchActuator.Set(ControlMode::PercentOutput, 0.0);
+    else {
+        if (m_hatchActuatorDeadband.Get() <= abs(m_actuatorBottomLimit.Get() - m_hatchActuator.GetSelectedSensorPosition())) {
```
You want to make sure that the difference between the sensor position and the limits is less than the deadband.

COREProgrammers committed:
Hopefully the last commit

njohnson0000 commented:
@Julia-Pockat I feel like this could be the last one.

Julia-Pockat commented on src/HatchIntakeSubsystem.cpp:
```diff
@@ -31,44 +31,24 @@ void HatchIntakeSubsystem::teleop() {
-//Actuates the intake, deadbanded.
-void HatchIntakeSubsystem::SetHatchActuator(bool goUp) {
+// isHatchIn is whether the hatch needs to be outtaked or intaked.
+void HatchIntakeSubsystem::SetIntake(bool setIntakeDirection, bool isHatchIn, bool goUp) {
```
Maybe you could make an enum for the intake directions instead of using a bool to make it more clear about which way you would be intaking for whatever is passed in.

Julia-Pockat commented on src/HatchIntakeSubsystem.cpp:
```diff
@@ -78,13 +58,28 @@ void HatchIntakeSubsystem::SetHatchActuator(bool goUp) {
     }
+
+    if(setIntakeDirection) {
+        if(isHatchIn) {
```
For this I think that you could just call the HatchIn() method in this class instead of adding an additional parameter for isHatch in.

Julia-Pockat commented on src/HatchIntakeSubsystem.cpp:
```diff
@@ -31,44 +31,24 @@ void HatchIntakeSubsystem::teleop() {
     if(!(operatorJoystick->GetButton(CORE::COREJoystick::JoystickButton::RIGHT_BUTTON) || operatorJoystick->GetButton(CORE::COREJoystick::JoystickButton::RIGHT_TRIGGER))){
-        SetHatchActuator(true);
+        SetIntake(true, !HatchIn(), true);
         m_hatchIntake.Set(ControlMode::PercentOutput, 0.0);
```
I don't think that you would want to set the hatch intake to 0 percent output right after you turn it on. This if statement should just stop running the motors when the button is stopped being pressed.

Julia-Pockat commented on src/HatchIntakeSubsystem.cpp:
```diff
@@ -31,44 +31,24 @@ void HatchIntakeSubsystem::teleop() {
 
-    // Actuates the intake back up when the button is no longer being pressed
+    // Actuates the intake back up when the button is no longer being pressed and stops the intake.
     if(!(operatorJoystick->GetButton(CORE::COREJoystick::JoystickButton::RIGHT_BUTTON) || operatorJoystick->GetButton(CORE::COREJoystick::JoystickButton::RIGHT_TRIGGER))){
```
Why do you have an or statement that checks to see if a certain button is not pressed? According to this, if the right button is not being pressed, then the if statement would be executed and the intake motors would be run.

Julia-Pockat commented on src/HatchIntakeSubsystem.cpp:
```diff
@@ -31,44 +31,24 @@ void HatchIntakeSubsystem::teleop() {
     if (operatorJoystick->GetButton(CORE::COREJoystick::JoystickButton::RIGHT_TRIGGER)) {
-        SetHatchActuator(false);
-        SetIntake(true, !HatchIn());
+        SetIntake(true, !HatchIn(), false);
```
Is the hatch intake parameter supposed to be using the '!' symbol? Right now, this would mean that the value that is passed in would be the opposite of the current state of the hatch.

COREProgrammers committed:
The last commit(fingers crossed).
Changed the booleans to enum list.
Eliminated the third parameter on the setIntake function.

njohnson0000 commented:
@Julia-Pockat I have done the enum list so hopefully everything else is correct.

njbackhaus committed:
Merge branch 'master' into HatchIntakeSubsystemCode
